tidy3d/plugins/design/design.py

```diff
@@ -128,52 +322,175 @@ def run_batch(
-            Can be converted to ``pandas.DataFrame`` with ``.to_dataframe()``.
+        new_self = self.updated_copy(path_dir=path_dir)
+        result = new_self.run(fn=fn_pre, fn_post=fn_post)
+
+        return result
+
+    def _get_run_count(self) -> int:
+        # For MonteCarlo, Random, RandomCustom
+        if hasattr(self.method, "num_points"):
+            num_points = self.method.num_points
+
+        # For Grid
+        elif isinstance(self.method, MethodGrid):
+            num_points = len(MethodGrid.sample(self.parameters))
+
+        # For BayOpt
+        elif isinstance(self.method, MethodBayOpt):
+            num_points = self.method.initial_iter + self.method.n_iter
+
+        # For GenAlg
+        elif isinstance(self.method, MethodGenAlg):
+            num_points = self.method.solutions_per_pop * self.method.n_generations
+
+        # For ParticleSwarm
+        elif isinstance(self.method, MethodParticleSwarm):
+            num_points = self.method.n_particles * self.method.n_iter
+
+        return num_points
+
+    def estimate_cost(self, fn_pre: Callable) -> float:
         """
+        Compute the maximum FlexCredit charge for the entire design space computation.
+        Require a pre function that should return a Simulation object, Batch object, or collection of either.
+        The pre function is called to estimate the cost - complicated pre functions may cause long runtimes.
+        """
+        # Get output fn_pre for paramters at the lowest span / default
+        arg_dict = {}
+        for param in self.parameters:
+            if hasattr(param, "span"):
+                arg_dict[param.name] = param.span[0]
+            else:  # For ParameterAny objects
+                arg_dict[param.name] = param.allowed_values[0]
 
-        # run the functions using the `method.run_batch`
-        fn_args, fn_values, task_ids, batch_data = self.method.run_batch(
-            parameters=self.parameters,
-            fn_pre=fn_pre,
-            fn_post=fn_post,
-            path_dir=path_dir,
-            **batch_kwargs,
-        )
+        # Compute fn_pre
+        pre_out = fn_pre(**arg_dict)
 
-        # store the pre and post functions in a single source code
-        fn_source_pre = self.get_fn_source(fn_pre)
-        fn_source_post = self.get_fn_source(fn_post)
-        fn_source = self._make_batch_fn_source(fn_source_pre, fn_source_post)
+        def _estimate_sim_cost(sim):
+            job = Job(simulation=sim, task_name="estimate_cost")
 
-        # package the result
-        return self._package_run_results(
-            fn_args=fn_args,
-            fn_values=fn_values,
-            fn_source=fn_source,
-            task_ids=task_ids,
-            batch_data=batch_data,
+            estimate = job.estimate_cost()
+            job.delete()  # Deleted as only a test with initial parameters
+
+            return estimate
+
+        if isinstance(pre_out, Simulation):
```

Comment:
consider making this a bit recursive to simplify things?

```py
def estimate_cost(pre_out):

    if isinstance(pre_out, Simulation):
         job = Job(simulation=pre_out)
         time.sleep(2.0) # not sure, maybe this will give the server enough time to compute it?
         return  estimate_cost(job)
    if isinstance(pre_out, (Job, Batch)):
         cost = pre_out.esimate_cost()
         pre_out.delete()
         return cost
    if isinstance(pre_out, dict):
        return estimate_cost(tuple(pre_out.values()))
    if isinstance(pre_out, (list, tuple)):
        return sum(estimate_cost(x) for x in pre_out)
    return 0.0
```
         



```

Reply:
I've written this like this because it follows the same None handling logic as `Batch.estimate_cost()`. 

There's no race condition on our servers, they seem to queue and time out cost estimates correctly so I don't think a sleep is necessary.

We've got to break the lists and dicts down somewhere into component simulations and batches to take advantage of existing `Job.estimate_cost` or `Batch.estimate_cost` methods.

We can't return 0 for failed estimates as wouldn't people then believe the simulation is free?

Reply:
ok, I see. maybe this doesn't properly handle the None condition.

Btw the return 0 was meant to handle the case where pre_out was not sim-like.

If what you have before works, we can keep that.

Reply:
Ah sorry, that makes more sense